py3status/modules/window.py

```diff
@@ -0,0 +1,204 @@
+# -*- coding: utf-8 -*-
+"""
+Display window properties (i.e. title, class, instance).
+
+Configuration parameters:
+    cache_timeout: refresh interval for i3-msg (default 1)
+    format: display format for this module (default '{title}')
+
+Format placeholders:
+    {class} window class
+    {instance} window instance
+    {title} window title
+
+Examples:
+```
+# show heart instead of empty title
+window_title {
+    format = '{title}|\u2665'
+}
+
+# specify a length
+window_title {
+    format = '[\?max_length=55 {title}]'
+}
+```
+
+@author shadowprince (counter), Anon1234 (async)
+@license Eclipse Public License (counter), BSD (async)
+
+SAMPLE OUTPUT
+{'full_text': u'business_plan_final_3a.doc'}
+"""
+
+try:
+    import i3ipc
+    from threading import Thread
+
+    IPC = "i3ipc"
+except Exception:
+    IPC = "i3msg"
+
+STRING_ERROR = "invalid ipc `{}`"
+
+
+class Ipc:
+    """
+    """
+
+    def __init__(self, parent):
+        self.parent = parent
+
+        # deprecations?
+        self.max_width = getattr(parent, "max_width", 120)
+        self.placeholders = dict.fromkeys(
+            self.parent.py3.get_placeholders_list(self.parent.format), ""
+        )
+        self.placeholders.pop("ipc", None)
+        self.setup(parent)
+
+    def set_window_properties(self):
+        pass
+
+    def trim_width(self, string):
+        if len(string) > self.max_width:
+            string = string[: self.max_width - 1] + u"…"
+        return string
+
+
+class I3ipc(Ipc):
+    """
+    """
+
+    def setup(self, parent):
+        self.parent.cache_timeout = self.parent.py3.CACHE_FOREVER
+        self.last_window_properties = self.placeholders
+
+        # deprecations?
+        self.always_show = getattr(parent, "always_show", False)
+        self.empty_title = getattr(parent, "empty_title", "")
+
+        t = Thread(target=self.start)
+        t.daemon = True
+        t.start()
+
+    def start(self):
+        i3 = i3ipc.Connection()
+        self.change_title(i3)
+        for event in ["workspace::focus", "window::close"]:
+            i3.on(event, self.clear_title)
+        for event in ["window::title", "window::focus", "binding"]:
+            i3.on(event, self.change_title)
+        i3.main()
+
+    def clear_title(self, i3, event):
+        focused = i3.get_tree().find_focused()
+        if not focused.window_class:
+            focused.window_title = self.empty_title
+        self.update(focused)
+
+    def change_title(self, i3, event=None):
+        focused = i3.get_tree().find_focused()
+
+        # do not hide title when it is already visible (e.g. stacked/tabbed layout)
+        if not self.always_show and (
+            focused.border == "normal"
+            or focused.type == "workspace"
+            or (
+                focused.parent.layout in ("stacked", "tabbed")
+                and len(focused.parent.nodes) > 1
+            )
+        ):
+            focused.window_title = self.empty_title
```

Comment:
I feel it's okay to keep things simple on our end and asking to install dedicated tools for a more complicated behavior. If it's easy to move this to the parent class - sure, otherwise let people install i3ipc.